src/components/outline-nav/utils/get-item-slugs.ts

```diff
@@ -0,0 +1,21 @@
+import { OutlineLinkItem } from '../types'
+
+/**
+ * Flatten an array of OutlineLinkItem[] into an array of #hash slugs
+ */
+function getItemSlugs(items: OutlineLinkItem[]) {
+	return items.reduce((acc: string[], item: OutlineLinkItem) => {
+		// Note that baseUrl isn't important, we're just using it to get the hash
+		const urlObject = new URL(item.url, 'https://www.example.com')
+		const slug = urlObject.hash.replace('#', '') // get hash without leading #
+		if (slug !== '') {
+			acc.push(slug)
+		}
+		if ('items' in item) {
+			acc = acc.concat(getItemSlugs(item.items))
+		}
+		return acc
+	}, [])
+}
```

Comment:
We are recursing here already - but agree, `reduce` can be a bit hard to read sometimes. I can switch this to a `for` loop, and split out the `#hash` extraction part for clarity.

The reason we don't reference raw data here (slash why the `#hash` logic is necessary) is that we want `OutlineNavWithActive` to accept items with `{ title, url }` rather than `{ title, slug }`, as mentioned in the other thread on `slug` vs `url` in `OutlineLink`. We also have a follow-up task 🎟 [Clean up headings-related OutlineNav uses, to move formatting closer to server](https://app.asana.com/0/0/1204024966433303/f) to move more of the logic closer to the server, however it'll be higher effort so I'm thinking of it as a future optimization. All new uses of `OutlineNav` (eg in Integrations) will ideally generate outline items directly (kind of using `getStaticProps` as the business logic layer, rather than a mix of `getStaticProps` and logic within components).

Reply:
✅ updated to use a `for` loop, and split out `getHash(url: string)` 👍 Let me know if that works!